### ISSUE TITLE
Parser: 3 ideas to improve

### ISSUE THREAD

**Issue by robertomatute** [mareczek/international-phone-number#29]
Hi,

Thanks for this amazing directive.  It saves me hours of work!!

Three details, I think will improve the directive.

1) in the second parser you do:
```javascript
     delete ctrl.$error['international-phone-number']
```
I think is important to call:
```javascript
            ctrl.$setValidity('international-phone-number', true);
```
to clear the $invalid state of the input.

2) In the second parser too, you are returning the value received.  I think will be better to return:
```javascript
return element.intlTelInput("getNumber")
```
To provide the entire international number.

3) To maintain the standard angular way of handling the validation, I think it will be better to return undefined if it not valid in the parser.  Something like:
```javascript
            if (validity){
              //return value;
              return element.intlTelInput("getNumber")
            }else{ 
              return undefined;
            }
```

I hope it helps,

Best Regard,

Roberto

**Comment by mareczek** [comment]
@robertomatute Hi!, thanks for the tips! Could you take a look at the code in tag 0.0.6 and let me know if it looks better?

**Comment by jasonhjohnson** [comment]
Hi,

Do you have a compiled js version with these fixes in?

Thanks,
Jason

**Comment by robertomatute** [comment]
Hi,

This is the directive I Implemented:


/***********/

directives.directive('internationalPhoneNumber2', ['$timeout',
function($timeout) {
    return {
      restrict: 'A',
      require: '^ngModel',
      scope: {
        ngModel: '=',
        defaultCountry: '@'
      },
      link: function(scope, element, attrs, ctrl) {
        var handleWhatsSupposedToBeAnArray, options, read, watchOnce;
        read = function() {
          return ctrl.$setViewValue(element.val());
        };
        handleWhatsSupposedToBeAnArray = function(value) {
          if (value instanceof Array) {
            return value;
          } else {
            return value.toString().replace(/[ ]/g, '').split(',');
          }
        };
        options = {
          autoFormat: true,
          autoHideDialCode: true,
          defaultCountry: '',
          nationalMode: true,
          numberType: '',
          onlyCountries: void 0,
          preferredCountries: ['ar', 'bo', 'ca', 'cl', 'co', 'cr', 'do',
'ec', 'mx', 'pa', 'pe', 'es','us', 've' ],
          responsiveDropdown: false,
          utilsScript: ""
        };
        angular.forEach(options, function(value, key) {
          var option;
          if (!(attrs.hasOwnProperty(key) &&
angular.isDefined(attrs[key]))) {
            return;
          }
          option = attrs[key];
          if (key === 'preferredCountries') {
            return options.preferredCountries =
handleWhatsSupposedToBeAnArray(option);
          } else if (key === 'onlyCountries') {
            return options.onlyCountries =
handleWhatsSupposedToBeAnArray(option);
          } else if (typeof value === "boolean") {
            return options[key] = option === "true";
          } else {
            return options[key] = option;
          }
        });
        watchOnce = scope.$watch('ngModel', function(newValue) {
          $timeout(function() {
            options.defaultCountry = scope.defaultCountry;
            if (newValue !== null && newValue !== void 0 && newValue !==
'') {
              element.val(newValue);
            }
            element.intlTelInput(options);
            if (!(attrs.skipUtilScriptDownload !== void 0 ||
options.utilsScript)) {
              element.intlTelInput('loadUtils',
'/bower_components/intl-tel-input/lib/libphonenumber/build/utils.js');
            }
            return watchOnce();
          }, 0, false);
        });
        ctrl.$parsers.push(function(value) {
          if (!value) {
            return value;
          }
          return value.replace(/[^\d]/g, '');
        });
        ctrl.$parsers.push(function(value) {
          var validity;
          if (value) {
            validity = element.intlTelInput("isValidNumber");
            ctrl.$setValidity('phone', validity);
            //ctrl.$setValidity('', validity);
            if (validity){
              //return value;
              return element.intlTelInput("getNumber")

            }else{
              return undefined;
            }
          } else {
            value = '';
            ctrl.$setValidity('phone', true);
            delete ctrl.$error['phone'];
          }
          return value;
        });
        element.on('blur keyup change', function(event) {
          return scope.$apply(read);
        });
        return element.on('$destroy', function() {
          return element.off('blur keyup change');
        });
      }
    };
  }]);

/***********/

Regards,

Roberto

**Comment by jasonhjohnson** [comment]
Thanks @robertomatute It seems like the form containing our international phone number field is marked as invalid on intial page load even if the number in that field is valid. We have to click inside the field and then the form is marked as valid. I console.logged inside the watch method and it looks like it's being hit twice (the first time through newValue is undefined and the second time through it has our phone number). Any ideas what's going on here?

**Comment by mareczek** [comment]
This is fixed in v0.0.8
